src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,217 @@
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
+begin
+  induction n with n hn,
+  { rw [nat.desc_fac_zero], nontriviality (fin 0 ↪ β),
+    obtain ⟨f, g, ne⟩ := exists_pair_ne (fin 0 ↪ β),
+    exfalso,
+    apply ne,
+    ext x,
+    exact x.elim0 },
+
+  let extend : (fin n → β) → β → fin n.succ → β := λ f b, fin.cons b f,
+
+  let equiv_classes : (fin n ↪ β) → finset (fin n.succ ↪ β) :=
+    λ f : fin n ↪ β, univ.filter (λ g : fin n.succ ↪ β, ∃ k : β, extend f k = g),
+
+  have mem_equiv : ∀ f g, g ∈ equiv_classes f ↔ ∃ k : β, extend f k = g, by simp [equiv_classes],
+
+  have all_injf_covered : univ = univ.bUnion equiv_classes,
+  { apply subset.antisymm,
+    { rintros f -,
+      rw mem_bUnion,
+      refine ⟨⟨fin.tail f, λ _ _ h, fin.succ_inj.mp $ f.injective h⟩, _⟩,
+      suffices : ∃ (a : β), extend (fin.tail ⇑f) a = ⇑f, by simpa,
+
+      use f 0,
+      simp [extend] },
+    { exact subset_univ _ } },
+
+  have equiv_class_size : ∀ f : fin n ↪ β, |equiv_classes f| = ‖β‖ - n,
+  {
+    intro f,
+
+    let poss_vals := univ \ finset.map ⟨f, f.inj'⟩ univ,
+    have num_poss_vals : |poss_vals| = ‖β‖ - n, by simp [poss_vals, card_univ, card_sdiff],
+    have mem_poss_vals : ∀ t, t ∈ poss_vals ↔ ∀ (x : fin n), ¬f x = t, by simp [poss_vals],
+
+    apply le_antisymm,
+    { by_contra card_too_big,
+      suffices : ∃ (x : fin n.succ ↪ β) (hx : x ∈ equiv_classes f)
+                   (y : fin n.succ ↪ β) (hy : y ∈ equiv_classes f), x ≠ y ∧ x 0 = y 0,
+      { obtain ⟨x, x_equiv, y, y_equiv, x_ne_y, x_y_agree⟩ := this,
+        rw [mem_equiv] at x_equiv y_equiv,
+        obtain ⟨x_zero, x_equiv⟩ := x_equiv,
+        obtain ⟨y_zero, y_equiv⟩ := y_equiv,
+        simp only [extend] at x_y_agree x_equiv y_equiv,
+
+        apply x_ne_y,
+        ext t,
+        revert t,
+        refine fin.induction _ _,
+          exact x_y_agree,
+        rintros t -,
+        rw [←x_equiv, ←y_equiv, fin.cons_succ, fin.cons_succ] },
+
+      apply finset.exists_ne_map_eq_of_card_lt_of_maps_to,
+      push_neg at card_too_big,
+      rw ←num_poss_vals at card_too_big,
+      exact card_too_big,
+
+      intros g g_equiv, rw mem_equiv at g_equiv,
+      obtain ⟨k, g_equiv⟩ := g_equiv,
+      simp only [mem_poss_vals],
+      simp only [extend] at g_equiv,
+
+      have : g 0 = k, by rw [←g_equiv, fin.cons_zero],
+
+      intros x eq,
+      have : g x.succ = g 0, by rw [←eq, ←g_equiv, fin.cons_succ],
+      apply fin.succ_ne_zero x,
+      exact g.injective this },
+
+    { let extended : finset (fin n.succ ↪ β) :=
+        finset.map ⟨λ x : {x // x ∈ poss_vals}, ⟨extend f x, _⟩, _⟩ poss_vals.attach,
+      rotate,
+      { intros a₁ a₂,
+        apply fin.induction_on a₁;
+        apply fin.induction_on a₂,
+        -- not sure how to do this `induction_on` using the `induction` tactic
+        -- (all variants of `induction using fin.something` didn't work for me)
```

Comment:
That's fine: `induction using` is pretty unreliable, and mathlib extensively just uses `apply` for this.